### PR TITLE
REF: handle dtype dispatch in libhashtable, share Vector/Hashtable code

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/hashtable_func_helper.pxi.in [hunk 1]
```diff
@@ -310,3 +310,163 @@ def mode_{{dtype}}(const {{dtype}}_t[:] values, bint dropna):
...
+
+    elif htfunc_t is float64_t:
+        return value_count_float64(values, dropna)
+    elif htfunc_t is float32_t:
+        return value_count_float32(values, dropna)
+
+    elif htfunc_t is complex128_t:
+        return value_count_complex128(values, dropna)
+    elif htfunc_t is complex64_t:
+        return value_count_complex64(values, dropna)
```
Comment (mzeitlin11) [inline pandas/_libs/hashtable_func_helper.pxi.in:361]: Can't wait for the arrival of pattern matching in python :)

**Comment by jreback** [comment]
can you rebase (havne't looked yet)

**Comment by jreback** [comment]
ok looks fine, any perf diffs? (maybe expect small marginal improvements)

**Comment by jbrockmendel** [comment]
> ok looks fine, any perf diffs? (maybe expect small marginal improvements)

i expect it to be a wash

**Review by jreback — APPROVED** [review]

**Pull request merged**